source/_posts/2018-01-02-Getting-started-with-Vapor-3.md

```diff
@@ -0,0 +1,344 @@
+    RouteGroup(cascadingTo: router).group("posts") { group in
+        let postController = PostController()
+    }
+}
+```
+
+And for now, let's just create an empty controller:
+
+```swift
+import Vapor
+
+final class PostController {
+  
+}
+```
+
+#### Retrieving all instances
+
+To return all posts created (e.g. by requesting `GET /posts`), we're going to query them like this:
+
+```swift
+func all(req: Request) throws -> Future<[Post]> {
+    return Post.query(on: req).all()
+}
+```
+
+There's a couple of things to notice in the above snippet when we compare it to Vapor 2. 
+
+First is that we're not specifying what connection to perform the query on. In this case, we're able to use the request itself because we made `Request` conform to `DatabaseConnectable`. This makes it very explicit for who is responsible for performing the lookup. Without having tried it yet, this should help us make tests as well since we should potentially be able mock the database connection and instead of hitting a database, we could return in-memory objects created in our tests.
+
+Next, there's the return type which is now `Future<[Post]>` instead of `[Post]` as you might have expected coming from Vapor 2. [Async, Streams, Futures and Reactive Programming](https://docs.vapor.codes/3.0/async/getting-started/) are central topics in Vapor 3 in order to increase the performance of the framework and it will change the way we work. Without going too much into details, one would of thinking of the concepts could be:
+
+- Futures: Values that at some point in time will exist. The future part is a wrapper that allows us to continue the work we want to do when the value arrives.
+- Streams: If a function returns one value when it's done executing, a stream returns an endless number of values. Think of it as generally always running which might return values once in a while.
+- Reactive Programming: Related to futures and streams are reactive programming. It's a pattern for dealing with these types of data, usually using functional programming. You can think of it as a way of transforming or dealing with data when it comes in through a future or a stream.
+
+If you've worked with any of the reactive programming frameworks, these concepts introduced in Vapor 3 should be familiar.
+
+#### Retriving one specific instance
+
+To return one specific post (e.g. by requesting `GET /posts/:id`), we're going to request that instance like this:
+
+```Swift
+func specific(req: Request) throws -> Future<Post> {
+    return try req.parameter(Post.self)
+}
+```
+
+This is very similar to what we've used to in Vapor 2 with the exception that fetching a parameter from the request now returns a `Future`.
+
+#### Creating a new instance
+
+To create a new post (e.g. by requesting `POST /posts`), we're going to do it like this:
+
+```swift
+func create(req: Request) throws -> Future<Post> {
+    let post = try req.content.decode(Post.self)
+    return post.save(on: req).transform(to: post)
```

Comment:
the `.transform(to: post)` is not necessary as save already returns a `Future<Post>`